File: backend/app/services/validation_service.py
from typing import Dict, Any, List, Optional, Callable
import re
import logging
from datetime import datetime
from app.core.config_loader import config_loader

logger = logging.getLogger("validation_service")


class ValidationRule:
    def __init__(self, rule_type: str, rule_value: Any, message: str):
        self.rule_type = rule_type
        self.rule_value = rule_value
        self.message = message


class ValidationResult:
    def __init__(self, valid: bool, errors: List[str] = None):
        self.valid = valid
        self.errors = errors or []


class ValidationEngine:
    _custom_rules: Dict[str, Callable] = {}
    
    @classmethod
    def register_custom_rule(cls, rule_type: str, validator: Callable):
        cls._custom_rules[rule_type] = validator
    
    @classmethod
    def validate_field(cls, field_value: Any, rules: List[Dict]) -> ValidationResult:
        errors = []
        
        for rule in rules:
            rule_type = rule.get("rule_type")
            rule_value = rule.get("rule_value")
            message = rule.get("message")
            
            if not rule_type:
                continue
            
            is_valid = cls._validate_single_rule(field_value, rule_type, rule_value)
            if not is_valid:
                errors.append(message)
        
        return ValidationResult(valid=len(errors) == 0, errors=errors)
    
    @classmethod
    def validate_form(cls, form_data: Dict[str, Any], fields: List[Dict]) -> ValidationResult:
        all_errors = []
        
        for field in fields:
            field_code = field.get("fieldCode")
            field_name = field.get("fieldName", field_code)
            required = field.get("required", False)
            rules = field.get("rules", [])
            
            field_value = form_data.get(field_code)
            
            if required and (field_value is None or field_value == ""):
                all_errors.append(f"{field_name} 不能为空")
                continue
            
            if field_value is not None and field_value != "":
                field_result = cls.validate_field(field_value, rules)
                if not field_result.valid:
                    all_errors.extend(field_result.errors)
        
        return ValidationResult(valid=len(all_errors) == 0, errors=all_errors)
    
    @classmethod
    def _validate_single_rule(cls, field_value: Any, rule_type: str, rule_value: Any) -> bool:
        if rule_type in cls._custom_rules:
            return cls._custom_rules[rule_type](field_value, rule_value)
        
        validators = {
            "minLength": cls._validate_min_length,
            "maxLength": cls._validate_max_length,
            "min": cls._validate_min,
            "max": cls._validate_max,
            "minimum": cls._validate_min,
            "maximum": cls._validate_max,
            "pattern": cls._validate_pattern,
            "email": cls._validate_email,
            "phone": cls._validate_phone,
            "idCard": cls._validate_id_card,
            "url": cls._validate_url,
            "enum": cls._validate_enum,
            "dateMin": cls._validate_date_min,
            "dateMax": cls._validate_date_max,
            "custom": cls._validate_custom_script,
        }
        
        validator = validators.get(rule_type)
        if validator:
            return validator(field_value, rule_value)
        
        return True
    
    @staticmethod
    def _validate_min_length(value: Any, min_len: int) -> bool:
        if value is None:
            return True
        return len(str(value)) >= min_len
    
    @staticmethod
    def _validate_max_length(value: Any, max_len: int) -> bool:
        if value is None:
            return True
        return len(str(value)) <= max_len
    
    @staticmethod
    def _validate_min(value: Any, min_val: float) -> bool:
        if value is None or value == "":
            return True
        try:
            return float(value) >= min_val
        except (ValueError, TypeError):
            return False
    
    @staticmethod
    def _validate_max(value: Any, max_val: float) -> bool:
        if value is None or value == "":
            return True
        try:
            return float(value) <= max_val
        except (ValueError, TypeError):
            return False
    
    @staticmethod
    def _validate_pattern(value: Any, pattern: str) -> bool:
        if value is None or value == "":
            return True
        try:
            return bool(re.match(pattern, str(value)))
        except re.error:
            return False
    
    @staticmethod
    def _validate_email(value: Any, _: Any = None) -> bool:
        if value is None or value == "":
            return True
        email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(email_pattern, str(value)))
    
    @staticmethod
    def _validate_phone(value: Any, _: Any = None) -> bool:
        if value is None or value == "":
            return True
        phone_pattern = r'^1[3-9]\d{9}$'
        return bool(re.match(phone_pattern, str(value)))
    
    @staticmethod
    def _validate_id_card(value: Any, _: Any = None) -> bool:
        if value is None or value == "":
            return True
        id_card_pattern = r'^[1-9]\d{5}(18|19|20)\d{2}(0[1-9]|1[0-2])(0[1-9]|[12]\d|3[01])\d{3}[\dXx]$'
        return bool(re.match(id_card_pattern, str(value)))
    
    @staticmethod
    def _validate_url(value: Any, _: Any = None) -> bool:
        if value is None or value == "":
            return True
        url_pattern = r'^https?://[^\s/$.?#].[^\s]*$'
        return bool(re.match(url_pattern, str(value)))
    
    @staticmethod
    def _validate_enum(value: Any, enum_values: List) -> bool:
        if value is None or value == "":
            return True
        return str(value) in [str(v) for v in enum_values]
    
    @staticmethod
    def _validate_date_min(value: Any, min_date: str) -> bool:
        if value is None or value == "":
            return True
        try:
            date_val = datetime.strptime(str(value), "%Y-%m-%d")
            min_dt = datetime.strptime(min_date, "%Y-%m-%d")
            return date_val >= min_dt
        except ValueError:
            return False
    
    @staticmethod
    def _validate_date_max(value: Any, max_date: str) -> bool:
        if value is None or value == "":
            return True
        try:
            date_val = datetime.strptime(str(value), "%Y-%m-%d")
            max_dt = datetime.strptime(max_date, "%Y-%m-%d")
            return date_val <= max_dt
        except ValueError:
            return False
    
    @staticmethod
    def _validate_custom_script(value: Any, script_config: Dict) -> bool:
        if value is None or value == "":
            return True
        
        script_type = script_config.get("type", "simple")
        
        if script_type == "simple":
            condition = script_config.get("condition", "")
            try:
                return eval(condition, {}, {"value": value})
            except Exception as e:
                logger.warning("[ValidationService] custom script eval 失败 condition=%s error=%s",
                               condition, e)
                return False
        
        elif script_type == "lambda":
            code = script_config.get("code", "")
            try:
                validator = eval(code)
                return validator(value)
            except Exception as e:
                logger.warning("[ValidationService] custom script lambda 失败 code=%s error=%s",
                               code, e)
                return False
        
        return True


validation_engine = ValidationEngine()

File: backend-app/src/main/java/com/sitech/prodai/service/ValidationService.java
package com.sitech.prodai.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * 表单校验引擎 —— 对齐 Python {@code app/services/validation_service.py::ValidationEngine}。
 *
 * <p>支持以下规则类型：
 * <ul>
 *   <li>minLength / maxLength：长度范围</li>
 *   <li>min / max / minimum / maximum：数值范围</li>
 *   <li>pattern：正则匹配</li>
 *   <li>email / phone / idCard / url：内置格式校验</li>
 *   <li>enum：枚举值</li>
 *   <li>dateMin / dateMax：日期范围（yyyy-MM-dd）</li>
 *   <li>custom：自定义脚本（Java 端只支持 simple 类型，lambda 类型返回 true）</li>
 * </ul>
 *
 * <p>注意：Python 版本支持 eval / lambda 自定义脚本，Java 出于安全考虑
 * 不支持任意脚本执行，仅对 simple 类型做有限的支持（基于 condition 字符串的简单比较）。
 */
@Service
public class ValidationService {

    private static final Logger log = LoggerFactory.getLogger(ValidationService.class);

    private static final Pattern EMAIL_PATTERN =
            Pattern.compile("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$");
    private static final Pattern PHONE_PATTERN =
            Pattern.compile("^1[3-9]\\d{9}$");
    private static final Pattern ID_CARD_PATTERN =
            Pattern.compile("^[1-9]\\d{5}(18|19|20)\\d{2}(0[1-9]|1[0-2])(0[1-9]|[12]\\d|3[01])\\d{3}[\\dXx]$");
    private static final Pattern URL_PATTERN =
            Pattern.compile("^https?://[^\\s/$.?#].[^\\s]*$");
    private static final DateTimeFormatter DATE_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd");

    /**
     * 校验单个字段 —— 对齐 Python validate_field。
     *
     * @param fieldValue 字段值
     * @param rules      规则列表，每项 {rule_type, rule_value, message}
     * @return 校验结果 {valid, errors}
     */
    public ValidationResult validateField(Object fieldValue, List<Map<String, Object>> rules) {
        List<String> errors = new ArrayList<>();
        if (rules == null) {
            return new ValidationResult(true, errors);
        }
        for (Map<String, Object> rule : rules) {
            String ruleType = str(rule.get("rule_type"));
            if (ruleType == null || ruleType.isEmpty()) {
                continue;
            }
            Object ruleValue = rule.get("rule_value");
            String message = str(rule.get("message"));
            if (!validateSingleRule(fieldValue, ruleType, ruleValue)) {
                errors.add(message);
            }
        }
        return new ValidationResult(errors.isEmpty(), errors);
    }

    /**
     * 校验整张表单 —— 对齐 Python validate_form。
     *
     * @param formData 表单数据 {field_code: value}
     * @param fields   字段定义列表，每项含 {fieldCode, fieldName, required, rules}
     * @return 校验结果 {valid, errors}
     */
    public ValidationResult validateForm(Map<String, Object> formData, List<Map<String, Object>> fields) {
        List<String> allErrors = new ArrayList<>();
        if (fields == null) {
            return new ValidationResult(true, allErrors);
        }
        for (Map<String, Object> field : fields) {
            String fieldCode = str(field.get("fieldCode"));
            String fieldName = firstNonBlank(str(field.get("fieldName")), fieldCode);
            boolean required = Boolean.TRUE.equals(field.get("required"));
            Object fieldValue = formData == null ? null : formData.get(fieldCode);

            if (required && isEmptyValue(fieldValue)) {
                allErrors.add(fieldName + " 不能为空");
                continue;
            }
            if (!isEmptyValue(fieldValue)) {
                Object rulesObj = field.get("rules");
                if (rulesObj instanceof List<?> rawRules) {
                    List<Map<String, Object>> rules = new ArrayList<>();
                    for (Object r : rawRules) {
                        if (r instanceof Map<?, ?> rm) {
                            @SuppressWarnings("unchecked")
                            Map<String, Object> cast = (Map<String, Object>) rm;
                            rules.add(cast);
                        }
                    }
                    ValidationResult r = validateField(fieldValue, rules);
                    allErrors.addAll(r.errors);
                }
            }
        }
        return new ValidationResult(allErrors.isEmpty(), allErrors);
    }

    // ==================== 单条规则校验 ====================

    @SuppressWarnings("unchecked")
    private boolean validateSingleRule(Object fieldValue, String ruleType, Object ruleValue) {
        switch (ruleType) {
            case "minLength":
                return validateMinLength(fieldValue, toInt(ruleValue, 0));
            case "maxLength":
                return validateMaxLength(fieldValue, toInt(ruleValue, Integer.MAX_VALUE));
            case "min", "minimum":
                return validateMin(fieldValue, toDouble(ruleValue, Double.NEGATIVE_INFINITY));
            case "max", "maximum":
                return validateMax(fieldValue, toDouble(ruleValue, Double.POSITIVE_INFINITY));
            case "pattern":
                return validatePattern(fieldValue, str(ruleValue));
            case "email":
                return validateEmail(fieldValue);
            case "phone":
                return validatePhone(fieldValue);
            case "idCard":
                return validateIdCard(fieldValue);
            case "url":
                return validateUrl(fieldValue);
            case "enum":
                return validateEnum(fieldValue, ruleValue instanceof List<?> list ? list : List.of());
            case "dateMin":
                return validateDateMin(fieldValue, str(ruleValue));
            case "dateMax":
                return validateDateMax(fieldValue, str(ruleValue));
            case "custom":
                if (ruleValue instanceof Map<?, ?> scriptConfig) {
                    return validateCustomScript(fieldValue, (Map<String, Object>) scriptConfig);
                }
                return true;
            default:
                return true;
        }
    }

    private boolean validateMinLength(Object value, int minLen) {
        if (value == null) {
            return true;
        }
        return String.valueOf(value).length() >= minLen;
    }

    private boolean validateMaxLength(Object value, int maxLen) {
        if (value == null) {
            return true;
        }
        return String.valueOf(value).length() <= maxLen;
    }

    private boolean validateMin(Object value, double minVal) {
        if (isEmptyValue(value)) {
            return true;
        }
        try {
            return Double.parseDouble(String.valueOf(value)) >= minVal;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private boolean validateMax(Object value, double maxVal) {
        if (isEmptyValue(value)) {
            return true;
        }
        try {
            return Double.parseDouble(String.valueOf(value)) <= maxVal;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private boolean validatePattern(Object value, String pattern) {
        if (isEmptyValue(value) || pattern == null || pattern.isEmpty()) {
            return true;
        }
        try {
            return Pattern.matches(pattern, String.valueOf(value));
        } catch (PatternSyntaxException e) {
            return false;
        }
    }

    private boolean validateEmail(Object value) {
        if (isEmptyValue(value)) {
            return true;
        }
        return EMAIL_PATTERN.matcher(String.valueOf(value)).matches();
    }

    private boolean validatePhone(Object value) {
        if (isEmptyValue(value)) {
            return true;
        }
        return PHONE_PATTERN.matcher(String.valueOf(value)).matches();
    }

    private boolean validateIdCard(Object value) {
        if (isEmptyValue(value)) {
            return true;
        }
        return ID_CARD_PATTERN.matcher(String.valueOf(value)).matches();
    }

    private boolean validateUrl(Object value) {
        if (isEmptyValue(value)) {
            return true;
        }
        return URL_PATTERN.matcher(String.valueOf(value)).matches();
    }

    private boolean validateEnum(Object value, List<?> enumValues) {
        if (isEmptyValue(value)) {
            return true;
        }
        String strValue = String.valueOf(value);
        for (Object v : enumValues) {
            if (v != null && strValue.equals(String.valueOf(v))) {
                return true;
            }
        }
        return false;
    }

    private boolean validateDateMin(Object value, String minDate) {
        if (isEmptyValue(value) || minDate == null || minDate.isEmpty()) {
            return true;
        }
        try {
            LocalDate dateVal = LocalDate.parse(String.valueOf(value), DATE_FMT);
            LocalDate min = LocalDate.parse(minDate, DATE_FMT);
            return !dateVal.isBefore(min);
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    private boolean validateDateMax(Object value, String maxDate) {
        if (isEmptyValue(value) || maxDate == null || maxDate.isEmpty()) {
            return true;
        }
        try {
            LocalDate dateVal = LocalDate.parse(String.valueOf(value), DATE_FMT);
            LocalDate max = LocalDate.parse(maxDate, DATE_FMT);
            return !dateVal.isAfter(max);
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    /**
     * 自定义脚本校验 —— 对齐 Python _validate_custom_script。
     *
     * <p>Java 端出于安全考虑不支持任意脚本执行，仅对 simple 类型做有限支持：
     * <ul>
     *   <li>支持 condition 形如 "value == 'foo'" / "value != 'bar'" / "value > 10" 等简单比较</li>
     *   <li>lambda 类型直接返回 true（不执行）</li>
     * </ul>
     */
    @SuppressWarnings("unchecked")
    private boolean validateCustomScript(Object value, Map<String, Object> scriptConfig) {
        if (isEmptyValue(value)) {
            return true;
        }
        String scriptType = str(scriptConfig.getOrDefault("type", "simple"), "simple");
        if ("simple".equals(scriptType)) {
            String condition = str(scriptConfig.get("condition"));
            return evaluateSimpleCondition(value, condition);
        }
        // lambda 类型在 Java 端不支持，直接返回 true
        return true;
    }

    /**
     * 简单条件评估 —— 仅支持 value == X / value != X / value > X / value < X / value >= X / value <= X
     * value 字符串本身不能包含引号以防止注入。
     */
    private boolean evaluateSimpleCondition(Object value, String condition) {
        if (condition == null || condition.isEmpty()) {
            return true;
        }
        try {
            String cond = condition.trim();
            // 仅支持以 "value" 开头的简单比较
            if (!cond.startsWith("value")) {
                log.warn("[ValidationService] 不支持的 condition: {}", cond);
                return false;
            }
            String rest = cond.substring(5).trim();
            String op = null;
            String operand = null;
            String[] ops = {"==", "!=", ">=", "<=", ">", "<"};
            for (String candidate : ops) {
                if (rest.startsWith(candidate)) {
                    op = candidate;
                    operand = rest.substring(candidate.length()).trim();
                    break;
                }
            }
            if (op == null || operand == null) {
                return false;
            }
            // 去掉引号
            if ((operand.startsWith("'") && operand.endsWith("'"))
                    || (operand.startsWith("\"") && operand.endsWith("\""))) {
                operand = operand.substring(1, operand.length() - 1);
            }
            String strValue = String.valueOf(value);
            switch (op) {
                case "==":
                    return strValue.equals(operand);
                case "!=":
                    return !strValue.equals(operand);
                case ">":
                    return Double.parseDouble(strValue) > Double.parseDouble(operand);
                case "<":
                    return Double.parseDouble(strValue) < Double.parseDouble(operand);
                case ">=":
                    return Double.parseDouble(strValue) >= Double.parseDouble(operand);
                case "<=":
                    return Double.parseDouble(strValue) <= Double.parseDouble(operand);
                default:
                    return false;
            }
        } catch (Exception e) {
            log.warn("[ValidationService] 简单条件评估失败 condition={} err={}", condition, e.getMessage());
            return false;
        }
    }

    // ==================== 工具方法 ====================

    private boolean isEmptyValue(Object value) {
        if (value == null) {
            return true;
        }
        if (value instanceof String s) {
            return s.isEmpty();
        }
        if (value instanceof List<?> list) {
            return list.isEmpty();
        }
        if (value instanceof Map<?, ?> map) {
            return map.isEmpty();
        }
        return false;
    }

    private int toInt(Object value, int defaultValue) {
        if (value instanceof Number n) {
            return n.intValue();
        }
        if (value == null) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(String.valueOf(value));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private double toDouble(Object value, double defaultValue) {
        if (value instanceof Number n) {
            return n.doubleValue();
        }
        if (value == null) {
            return defaultValue;
        }
        try {
            return Double.parseDouble(String.valueOf(value));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value);
    }

    private String str(Object value, String defaultValue) {
        if (value == null) {
            return defaultValue;
        }
        String s = String.valueOf(value);
        return s.isEmpty() ? defaultValue : s;
    }

    private String firstNonBlank(String... values) {
        if (values == null) {
            return null;
        }
        for (String v : values) {
            if (v != null && !v.isBlank()) {
                return v;
            }
        }
        return null;
    }

    /**
     * 校验结果 —— 对齐 Python ValidationResult。
     */
    public static class ValidationResult {
        public final boolean valid;
        public final List<String> errors;

        public ValidationResult(boolean valid, List<String> errors) {
            this.valid = valid;
            this.errors = errors == null ? List.of() : errors;
        }

        public Map<String, Object> toMap() {
            Map<String, Object> body = new LinkedHashMap<>();
            body.put("valid", valid);
            body.put("errors", errors);
            return body;
        }
    }
}
